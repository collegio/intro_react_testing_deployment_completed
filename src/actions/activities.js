import uuid from 'uuid';

export const addActivity = ({ name, activity_type = 'running', distance = '0 km' } = {}) => ({
    type: 'ADD_ACTIVITY',
    activity: {
        id: uuid(),
        name,
        activity_type,
        distance
    }
})
  
export const removeActivity = ({ id } = {}) => ({
    type: 'REMOVE_ACTIVITY',
    id
});

export const editActivity = (id, updates) => ({
    type: 'EDIT_ACTIVITY',
    id,
    updates
});