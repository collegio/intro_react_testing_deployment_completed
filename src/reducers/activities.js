// the default list of activities, which is empty
const activityDefaultState = [];

export default (state = activityDefaultState, action) => {
    switch (action.type) {

        case 'ADD_ACTIVITY':
            return [
                ...state,
                action.activity
            ];

        case 'REMOVE_ACTIVITY':
            return state.filter(({ id }) => id !== action.id);

        case 'EDIT_ACTIVITY':
            return state.map((activity) => {
                if (activity.id === action.id) {
                    return {
                        ...activity,
                        ...action.updates
                    };
                }
                else {
                    return activity;
                }
            });

        default:
            return state;
    }
};