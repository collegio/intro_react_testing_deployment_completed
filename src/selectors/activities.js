// Get visible activities

export default (activities, { text, activity_type, sort_by }) => {
    return activities.filter((activity) => {

        // match the text search
        const textNameMatch = activity.name.toLowerCase().includes(text.toLowerCase());

        // match the sport type
        let typeMatch = false;
        if (activity_type == 'all' || activity_type == activity.activity_type) {
            typeMatch = true;
        }

        return (textNameMatch && typeMatch);

    }).sort((a, b) => {

        if (sort_by === 'name') {
            return a.name > b.name ? 1 : -1;
        } else if (sort_by === 'distance') {

            // sort in reverse
            return a.distance > b.distance ? -1 : 1;
        }
    });
};