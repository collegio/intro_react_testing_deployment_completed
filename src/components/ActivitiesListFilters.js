import React from 'react';
import { connect } from 'react-redux';
import { setFilterText, setFilterType, sortByName, sortByDistance } from '../actions/filters';

export class ActivitiesListFilters extends React.Component {

    constructor(props) {
        super(props);

        this.onTextChange = this.onTextChange.bind(this);
        this.onSortChange = this.onSortChange.bind(this);
        this.onActivityTypeChange = this.onActivityTypeChange.bind(this);
    }

    onTextChange(e) {
        this.props.setFilterText(e.target.value);
    }

    onSortChange(e) {
        if (e.target.value === "name") {
            this.props.sortByName();
        }
        else if (e.target.value === "distance") {
            this.props.sortByDistance();
        }
    }

    onActivityTypeChange(e) {
        this.props.setFilterType(e.target.value);
    }

    render() {
        return (
            <div className="activities-list-filters">
                <h2>Filter Activities</h2>
                <div className="filter-group">
                    <label>Search For:</label><br />
                    <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
                </div>
                <div className="filter-group">
                    <label>Sort By:</label><br />
                    <select onChange={this.onSortChange}>
                        <option value="name">Name</option>
                        <option value="distance">Distance</option>
                    </select>
                </div>
                <div className="filter-group">
                    <label>Sport Type:</label><br />
                    <select onChange={this.onActivityTypeChange}>
                        <option value="all">All</option>
                        <option value="running">Running</option>
                        <option value="walking">Walking</option>
                        <option value="cycling">Cycling</option>
                        <option value="swimming">Swimming</option>
                    </select>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        setFilterText: (text) => dispatch(setFilterText(text)),
        setFilterType: (activity_type) => dispatch(setFilterType(activity_type)),
        sortByName: () => dispatch(sortByName()),
        sortByDistance: () => dispatch(sortByDistance()),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(ActivitiesListFilters);