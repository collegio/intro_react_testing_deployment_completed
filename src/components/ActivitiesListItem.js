import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { removeActivity } from '../actions/activities';

export const ActivitiesListItem = ({ dispatch, id, name, activity_type, distance }) => (
    <div className="activity">
        <div className="activity-info">
            <p>
                <Link to={'/edit/'+id}>
                    {name}<br />
                </Link>
                {activity_type} - {distance}
            </p>
            <button className="button button-clear" onClick={() => {
                dispatch(removeActivity({ id }));
            }}>Remove</button>
        </div>
    </div>
);

export default connect()(ActivitiesListItem);