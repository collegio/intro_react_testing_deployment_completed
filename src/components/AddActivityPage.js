import React from 'react';
import { connect } from 'react-redux';
import ActivityForm from './ActivityForm';
import { addActivity } from '../actions/activities';

export class AddActivityPage extends React.Component {

    constructor(props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
    }

    onSubmit(activity) {
        this.props.addActivity(activity);
        this.props.history.push('/');
    }

    render() {
        return (
            <div className="container">
                <h1>Add a Activity</h1>
                <ActivityForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addActivity: (activity) => dispatch(addActivity(activity))
});

export default connect(undefined, mapDispatchToProps)(AddActivityPage);