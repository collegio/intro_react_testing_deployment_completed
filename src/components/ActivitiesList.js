import React from 'react';
import { connect } from 'react-redux';
import ActivitiesListItem from './ActivitiesListItem';
import selectActivities from '../selectors/activities';

export const ActivitiesList = (props) => (
    <div className="activities-list">
        <h1>Activities List</h1>
        <div className="activity-list">
            {props.activities.map((activity) => {
                return <ActivitiesListItem key={activity.id} {...activity} />;
            })}
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        activities: selectActivities(state.activities, state.filters)
    }
};

export default connect(mapStateToProps)(ActivitiesList);