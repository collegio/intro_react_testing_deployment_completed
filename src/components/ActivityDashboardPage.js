import React from 'react';
import ActivitiesList from './ActivitiesList';
import ActivitiesListFilters from './ActivitiesListFilters';

export const ActivityDashboardPage = (props) => (
    <div className="container">
        <ActivitiesListFilters />
        <ActivitiesList />
    </div>
);

export default ActivityDashboardPage;