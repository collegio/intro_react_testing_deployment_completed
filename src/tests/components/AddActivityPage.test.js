import React from 'react';
import { shallow } from 'enzyme';
import { AddActivityPage } from '../../components/AddActivityPage';
import activities from '../fixtures/activities';

let addActivity, history, wrapper;

beforeEach(() => {
    addActivity = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<AddActivityPage addActivity={addActivity} history={history} />);
});

test('should render the AddActivityPage', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {  

    wrapper.find('ActivityForm').prop('onSubmit')(activities[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(addActivity).toHaveBeenLastCalledWith(activities[1]);
});