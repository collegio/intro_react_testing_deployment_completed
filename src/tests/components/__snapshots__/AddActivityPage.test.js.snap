// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the AddActivityPage 1`] = `
<div
  className="container"
>
  <h1>
    Add a Activity
  </h1>
  <ActivityForm
    onSubmit={[Function]}
  />
</div>
`;
